### DISCUSSION TITLE
Engine of volume pvc-29de027 dead unexpectedly, reattach the volume

### DISCUSSION THREAD

**Discussion by usma0118** [longhorn/longhorn#5223] [Q&A]
Environment: Ubuntu 22.10
Kubernetes: Microk8s: v1.24.3

Few volumes are not recovering after node restart.

## Log

`time="2023-01-06T11:59:02Z" level=warning msg="pvc-29de0274-5e00-4b45-8ea0-13223ba3d57c-e-5a640500: time=\"2023-01-06T11:16:25Z\" level=info msg=\"Adding backend: tcp://10.1.0.102:10195\""
time="2023-01-06T11:59:02Z" level=warning msg="pvc-29de0274-5e00-4b45-8ea0-13223ba3d57c-e-5a640500: 2023/01/06 11:16:25 cannot create an available backend for the engine from the addresses [tcp://10.1.0.102:10195]"
time="2023-01-06T11:59:02Z" level=warning msg="pvc-29de0274-5e00-4b45-8ea0-13223ba3d57c-e-5a640500: time=\"2023-01-06T11:16:37Z\" level=info msg=\"Starting with replicas [\\\"tcp://10.1.0.102:10120\\\"]\""
time="2023-01-06T11:59:02Z" level=warning msg="pvc-29de0274-5e00-4b45-8ea0-13223ba3d57c-e-5a640500: time=\"2023-01-06T11:16:37Z\" level=info msg=\"Connecting to remote: 10.1.0.102:10120\""
time="2023-01-06T11:59:02Z" level=warning msg="pvc-29de0274-5e00-4b45-8ea0-13223ba3d57c-e-5a640500: time=\"2023-01-06T11:16:37Z\" level=info msg=\"Opening: 10.1.0.102:10120\""
time="2023-01-06T11:59:02Z" level=warning msg="pvc-29de0274-5e00-4b45-8ea0-13223ba3d57c-e-5a640500: time=\"2023-01-06T11:16:37Z\" level=warning msg=\"backend tcp://10.1.0.102:10120 size does not match 3221225472 != 5368709120 in the engine initiation phase\""
time="2023-01-06T11:59:02Z" level=warning msg="pvc-29de0274-5e00-4b45-8ea0-13223ba3d57c-e-5a640500: time=\"2023-01-06T11:16:37Z\" level=info msg=\"Adding backend: tcp://10.1.0.102:10120\""
time="2023-01-06T11:59:02Z" level=warning msg="pvc-29de0274-5e00-4b45-8ea0-13223ba3d57c-e-5a640500: 2023/01/06 11:16:37 cannot create an available backend for the engine from the addresses [tcp://10.1.0.102:10120]"
time="2023-01-06T11:59:02Z" level=warning msg="pvc-29de0274-5e00-4b45-8ea0-13223ba3d57c-e-5a640500: time=\"2023-01-06T11:16:50Z\" level=info msg=\"Starting with replicas [\\\"tcp://10.1.0.102:10135\\\"]\""
time="2023-01-06T11:59:02Z" level=warning msg="pvc-29de0274-5e00-4b45-8ea0-13223ba3d57c-e-5a640500: time=\"2023-01-06T11:16:50Z\" level=info msg=\"Connecting to remote: 10.1.0.102:10135\""
time="2023-01-06T11:59:02Z" level=warning msg="pvc-29de0274-5e00-4b45-8ea0-13223ba3d57c-e-5a640500: time=\"2023-01-06T11:16:50Z\" level=info msg=\"Opening: 10.1.0.102:10135\""
time="2023-01-06T11:59:02Z" level=warning msg="pvc-29de0274-5e00-4b45-8ea0-13223ba3d57c-e-5a640500: time=\"2023-01-06T11:16:50Z\" level=warning msg=\"backend tcp://10.1.0.102:10135 size does not match 3221225472 != 5368709120 in the engine initiation phase\""
`

**Comment by innobead** [comment]
`backend tcp://10.1.0.102:10135 size does not match 3221225472 != 5368709120 in the engine initiation phase`

You should encounter this issue, https://github.com/longhorn/longhorn/issues/3601. What version you are using? This issue has been fixed in 1.3.2 and 1.4.0.

**Reply by usma0118** [reply]
version: 1.4.0

**Reply by usma0118** [reply]
Found: https://microk8s.io/docs/strict-confinement

**Reply by innobead** [reply]
Please provide the support bundle to have more context around it. Also. I would suggest creating a bug instead by https://github.com/longhorn/longhorn/issues/new?assignees=&labels=kind%2Fbug&template=bug_report.md&title=%5BBUG%5D and providing the related info. Thanks. 

cc @derekbit

**Reply by innobead** [reply]
oh, does the volume only have one replica??

**Reply by usma0118** [reply]
unfortunately Yes.

**Reply by derekbit** [reply]
How did the single replica volume encounter the issue? Did you change the number of replicas, encounter expansion failure or any steps?

**Reply by derekbit** [reply]
BTW, can you show the `ls -al ${disk-path}/replica/${volume}/` and `cat ${disk-path}/replica/${volume}/volume.meta`?
See if we can fix it by manual.

**Reply by usma0118** [reply]
> How did the single replica volume encounter the issue? Did you change the number of replicas, encounter expansion failure or any steps?

I have about 30 volumes running on this cluster. 4 of them have started behaving this way, after node restart. (not sure about root cause though).

**Reply by usma0118** [reply]
> BTW, can you show the `ls -al ${disk-path}/replica/${volume}/` and `cat ${disk-path}/replica/${volume}/volume.meta`? See if we can fix it by manual.
![image](https://user-images.githubusercontent.com/355154/211041738-6e22e7de-cba9-4167-8fba-f5c69c3ece7c.png)

`{"Size":5368709120,"Head":"volume-head-004.img","Dirty":true,"Rebuilding":false,"Error":"","Parent":"volume-snap-expand-5368709120.img","SectorSize":512,"BackingFilePath":""}`

**Reply by derekbit** [reply]
Got it. It seems the node was restarted whilst the volume was expanding.
@shuo-wu Any idea how to fix the volume?

**Reply by usma0118** [reply]
@shuo-wu any idea?

**Reply by shuo-wu** [reply]
It seems that the only replica (all replicas) failed to expand the size when the volume/node was restarted... Then when users try to re-attach the volume, the engine won't do expansion at the starting stage. As contrast, the rebuilding part will [check and re-try the expansion](https://github.com/longhorn/longhorn-engine/blob/7c550e728e9be6c30c8111f3c3ae660825b24452/pkg/sync/sync.go#L407-L409) before adding the replica to the engine. We may need a similar logic for the engine starting.


As for the workaround, can you cancel the expansion for this kind of volume by changing `v.Spec.Size` to `engine.Status.CurrentSize`? But it's better to provide the engine and volume CR before trying that.
```
kubectl -n longhorn-system get lhv <volume name>
kubectl -n longhorn-system get lhe <volume engine name>
```
